code/datums/traits.dm

```diff
@@ -0,0 +1,75 @@
+/datum/proc/add_trait(trait, source)
+	LAZYINITLIST(status_traits)
+
+	if(!status_traits[trait])
+		status_traits[trait] = list(source)
+	else
+		status_traits[trait] |= list(source)
+
+
+/datum/proc/remove_trait(trait, list/sources, force)
+	if(!status_traits)
+		return //nothing to remove
+
+	if(!status_traits[trait])
+		return
+
+	if(locate(ROUNDSTART_TRAIT) in status_traits[trait] && !force) //mob traits applied through roundstart cannot normally be removed
+		return
+
+	if(!sources) // No defined source cures the trait entirely.
+		status_traits -= trait
+		return
+
+	if(!islist(sources))
+		sources = list(sources)
+
+	if(LAZYLEN(sources))
+		for(var/S in sources)
+			if(S in status_traits[trait])
+				status_traits[trait] -= S
+	else
+		status_traits[trait] = list()
+
+	if(!LAZYLEN(status_traits[trait]))
+		status_traits -= trait
+
+/datum/proc/has_trait(trait, list/sources)
+	if(!status_traits)
+		return FALSE //well of course it doesn't have the trait
+
+	if(!status_traits[trait])
+		return FALSE
+
+	. = FALSE
```

Comment:
odd choice here due to it not being used anywhere else. I think it's more readable to have a return FALSE at the end.